bcipy/gui/gui_main.py

```diff
@@ -54,6 +70,268 @@ def get_id(self):
 
         return self.id
 
+class ComboBox(QComboBox):
+    """ComboBox with the same interface as a QLineEdit for getting and setting values.
+
+    Parameters:
+    -----------
+      options - list of items to appear in the lookup
+      selected_value - selected item; if this is not one of the options and new option will be
+        created for this value.
+    """
+
+    def __init__(self, options: List[str], selected_value: str, **kwargs):
+        super(ComboBox, self).__init__(**kwargs)
+        self.options = options
+        self.addItems(self.options)
+        self.setText(selected_value)
+        self.setEditable(True)
+
+    def setText(self, value: str):
+        """Sets the current index to the given value. If the value is not in the list of
+        options it will be added."""
+        if value not in self.options:
+            self.options = [value] + self.options
+            self.clear()
+            self.addItems(self.options)
+        self.setCurrentIndex(self.options.index(value))
+
+    def text(self):
+        """Gets the currentText."""
+        return self.currentText()
+
+
+class FormInput(QWidget):
+    """A form element with a label, help tip, and input control. The default control is a text
+    input. This object may be subclassed to specialize the input control or the arrangement of
+    widgets. FormInputs have the ability to show and hide themselves.
+
+    Parameters:
+    ----------
+      parameter - the option which will be configured using this input. Used to populate the label,
+        help text, and relevant options.
+      help_font_size - font size for the help text.
+      help_color - color of the help text.
+    """
+
+    def __init__(self,
+                 parameter: Parameter,
```

Comment:
I don't think this should be parameter specific. Same with the functions inheriting it below. The `SearchInput` is generic and should be here for certain. 

Let's say I want to use the boolean input to query the user for something other than a parameter (ex. an experiment field). I do think it can be usefully abstracted and reused. I also don't want to extend your PR. I'd be okay leaving them in here or with the params_form and addressing with the experiment tickets. You call!

Reply:
Yeah, we should discuss in planning. Alternatively experiment fields could use the same format as a Parameter. We would just need to add a `required` boolean property.